File: sample/ruby/map_util.rb
require 'csv'

$column_num, $row_num = 257, 257

def get_map_data_from_csv(csv_file, height_scale)
  # csvファイルから地図データを読み込み
  heights = []

  CSV.foreach("../map_file/#{csv_file}") do |row|
    for h in row
      h = h.to_f
      h = h != 0 ? (h * height_scale).floor : -1
      heights << h
    end
  end
#   puts "heights: #{heights}"

  max_height = heights.max.floor
  box_positions = (0...$row_num).map { |i| heights[i * $column_num, $column_num] }
  puts "max_height: #{max_height}"
#   puts "box_positions: #{box_positions}"
  { 'boxes' => box_positions, 'max_height' => max_height }
end

def get_box_color(height, max_height, high_color, low_color)
  # 高さによって色を変える
  r = (high_color[0] - low_color[0]) * height * 1.0 / max_height + low_color[0]
  g = (high_color[1] - low_color[1]) * height * 1.0 / max_height + low_color[1]
  b = (high_color[2] - low_color[2]) * height * 1.0 / max_height + low_color[2]

  [r, g, b]
end

File: sample/javaScript/mapUtil.mjs
import fs from 'fs';

const columnNum = 257;
const rowNum = 257;

export function getMapDataFromCSV(csvFile, heightScale) {
  const data = fs.readFileSync(`../map_file/${csvFile}`, 'utf8');
  const lines = data.split('\n');

  const heights = lines[0].split(',').map(h => {
    if (h !== '0') {
      return Math.floor(parseFloat(h) * heightScale);
    } else {
      return -1;
    }
  });

  const maxHeight = Math.floor(Math.max(...heights));
  console.log('max', maxHeight);

  const boxPositions = [];
  for (let i = 0; i < rowNum; i++) {
    const row = [];
    for (let j = 0; j < columnNum; j++) {
      const index = j + columnNum * i;
      row.push(heights[index]);
    }
    boxPositions.push(row);
  }

  const mapData = { boxes: boxPositions, maxHeight: maxHeight };
  return mapData;
}

export function getBoxColor(height, maxHeight, highColor, lowColor) {
  const r = (highColor[0] - lowColor[0]) * height / maxHeight + lowColor[0];
  const g = (highColor[1] - lowColor[1]) * height / maxHeight + lowColor[1];
  const b = (highColor[2] - lowColor[2]) * height / maxHeight + lowColor[2];

  return [r, g, b];
}
